filebeat/crawler/crawler.go

```diff
@@ -23,27 +24,44 @@ import (
 */
 
 type Crawler struct {
-	// Registrar object to persist the state
-	Registrar   *Registrar
-	prospectors []*prospector.Prospector
-	wg          sync.WaitGroup
+	registrar         *registrar.Registrar
+	prospectors       []*prospector.Prospector
+	wg                sync.WaitGroup
+	spooler           *spooler.Spooler
+	prospectorConfigs []*common.Config
 }
 
-func (c *Crawler) Start(prospectorConfigs []*common.Config, eventChan chan *input.FileEvent) error {
+func New(registrar *registrar.Registrar, spooler *spooler.Spooler, prospectorConfigs []*common.Config) (*Crawler, error) {
 
 	if len(prospectorConfigs) == 0 {
-		return fmt.Errorf("No prospectors defined. You must have at least one prospector defined in the config file.")
+		return nil, fmt.Errorf("No prospectors defined. You must have at least one prospector defined in the config file.")
 	}
 
-	logp.Info("Loading Prospectors: %v", len(prospectorConfigs))
+	return &Crawler{
+		registrar:         registrar,
+		spooler:           spooler,
+		prospectorConfigs: prospectorConfigs,
+	}, nil
+}
+
+func (c *Crawler) Start() error {
+
+	logp.Info("Loading Prospectors: %v", len(c.prospectorConfigs))
+
+	// Load the previous log file locations now, for use in prospector
+	err := c.registrar.LoadState()
+	if err != nil {
+		logp.Err("Error loading state: %v", err)
+		return err
+	}
 
 	// Get existing states
-	states := *c.Registrar.state
+	states := *c.registrar.GetStates()
```

Comment:
states only required to startup prosepctors? Registrar only required in order to call GetStates on startup? How about passing states as argument to Start instead of passing registrar to `crawler.New` ?